### PR TITLE
Add unit tests for CUDAMemoryManager, AllocatedMemory class

### PR DESCRIPTION
**Pull request by GuanLuo**
Also reorganize the location of utility function, and behavior of CUDAMemoryManager.

### CODE REVIEW

File: src/test/CMakeLists.txt [hunk 1]
```diff
@@ -43,3 +43,97 @@ install(
...
+)
+set_target_properties(
+  memory_test
+  PROPERTIES
+    SKIP_BUILD_RPATH TRUE
+    BUILD_WITH_INSTALL_RPATH TRUE
+    INSTALL_RPATH_USE_LINK_PATH FALSE
+    INSTALL_RPATH ""
+)
+target_include_directories(memory_test PRIVATE ${GTEST_INCLUDE_DIR} PRIVATE ${CUDA_INCLUDE_DIRS} PRIVATE ${CNMEM_PATH}/include)
```
Comment (deadeyegoodwin) [inline src/test/CMakeLists.txt]: break long line

File: src/test/memory_test.cc [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+}
+
+}  // namespace
+
+int
+main(int argc, char** argv)
+{
+  ::testing::InitGoogleTest(&argc, argv);
+  return RUN_ALL_TESTS();
+}
```
Comment (deadeyegoodwin) [inline src/test/memory_test.cc]: Put newline at end of file

File: src/core/server.cc [hunk 3]
```diff
@@ -149,9 +149,10 @@ InferenceServer::Init()
   CudaMemoryManager::Options cuda_options(
       min_supported_compute_capability_, cuda_memory_pool_size_);
   status = CudaMemoryManager::Create(cuda_options);
+  // If CUDA memory manager can't be created, just log error as the
+  // server can still function properly
   if (!status.IsOk()) {
-    ready_state_ = ServerReadyState::SERVER_FAILED_TO_INITIALIZE;
-    return status;
+    LOG_ERROR << status.Message();
```
Comment (deadeyegoodwin) [inline src/core/server.cc]: Don't you need to do something special to fail appropriately if load/unload is called and the manager didn't create correctly?

Comment (GuanLuo) [reply src/core/server.cc]: The Alloc / Free functions are checking if the CUDA memory manager is created, if not, the functions will return error

**Pull request merged**